---
title: 'Isoform Project'
output:
  prettydoc::html_pretty:
    theme: cayman
    toc: true
    number_sections: true
---

# Set Up

## Load Packages
```{r message=FALSE, warning=FALSE}
library(Gviz)
library(rtracklayer)
library(Rsamtools)
library(GenomicFeatures)
library(GenomicRanges)
library(rtracklayer)
library(Rsamtools)
library(GenomicAlignments)
library(VariantAnnotation)
library(tidyverse)
library(Rsubread)
```

## Randomly Sample from Cell Subclasses
The function below will take a directory of downloaded FASTQ files (e.g. scRNAseq samples from VISp region) and create a CSV containing a random sampling of 100 file names from a certain subclass (e.g. Pvalb)

* The CSV file created should be used to specify which FASTQ files to align in STAR script (which will create BAM files)

**Parameters:**

* **filedir:** the directory where FASTQs are found (assumed to contain only files for the specified region of interest)
* **outputcsv:** the directory and desired name for random sample, e.g. "../sample.csv" (*must end with .csv*)
* **metadatacsv:** path to CSV file containing metadata for each FASTQ (should have been downloaded with FASTQ files)
* **subclass:** as a character, subclass name (defined in metadatacsv)
* **region:** as a character, region name (defined in metadatacsv)

```{r}
makerandomsample <- function(metadatacsv, filedir, subclass, region, outputcsv) {
  # Get metadata for all files
  metadata <- read_csv(metadatacsv)
  
  # Get full list of downloaded FASTQ files in region of interest
  downloaded_files <- list.files(filedir, pattern="*.fastq.tar")
  downloaded_files <- downloaded_files %>% 
    str_remove(".fastq.tar") # Removing the extension to match metadata
  
  # Of the cells in the given subclass in the given region which were downloaded, randomly sample 100
  set.seed(12345) # For reproducibility
  metadata_sample <- metadata %>% 
    filter(subclass_label == subclass) %>% 
    filter(region_label == region) %>% 
    filter(exp_component_name %in% downloaded_files) %>% 
    select(exp_component_name) %>% 
    sample_n(100)
  
  # Save the random sample to a CSV
  metadata_sample %>% 
    write_csv(outputcsv, col_names=FALSE)
}
```

### Example Usage
```{r eval = FALSE}
makerandomsample(metadatacsv="/external/rprshnas01/netdata_kcni/stlab/Public/AIBS_scRNAseq_2019/mouse/2020_10_24/metadata.csv",
                 filedir="/external/rprshnas01/netdata_kcni/stlab/Public/AIBS_scRNAseq_2019/mouse/Raw/VISp",
                 subclass="L2/3 IT CTX-1", region="VISp", 
                 outputcsv="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Random100_PVALBvsL23_inVisp/l23_sample.csv")

makerandomsample(metadatacsv="/external/rprshnas01/netdata_kcni/stlab/Public/AIBS_scRNAseq_2019/mouse/2020_10_24/metadata.csv",
                 filedir="/external/rprshnas01/netdata_kcni/stlab/Public/AIBS_scRNAseq_2019/mouse/Raw/VISp",
                 subclass="Pvalb", region="VISp", 
                 outputcsv="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Random100_PVALBvsL23_inVisp/pvalb_sample.csv")
```

## Create Subclass-Level BAMs
The function below will create and index a merged BAM for a specific subclass

**Parameters:**

* **inputdir:** the directory where BAMs are found (i.e., the coord_bams directory produced from STAR script)
* **outputname:** the directory and desired name for new merged BAM, e.g. "../merged.bam" (*must end with .bam*)
* **chrID:** as a character, representing chromosome for selected reference, e.g. "NC_000082.7" for RefSeq OR "16" for Ensembl
* **coords:** as a vector, where the 1st value is the start coordinate and the 2nd is the end, e.g. c(0, 100)—note that unit is in bases, so 18,050kb becomes 18050000

```{r}
makesubclassBAM <- function(inputdir, outputname, chrID, coords) {
  
  # Create list of file paths for the BAMs
  bamfiles <- list.files(inputdir)
  bampaths <- paste0(inputdir, "/", bamfiles)
  
  # Index each BAM file (removing any previously created index files beforehand)
  bampaths_nobai <- str_subset(bampaths, ".bai", negate=TRUE)
  for (file in bampaths_nobai) {
    indexBam(file)
  }
  
  # Merge all BAMs into one which only contains gene of interest
  merged <- mergeBam(files=bampaths, destination=outputdir, 
                     overwrite=TRUE, region = GRanges(chrID, IRanges(coords[1], coords[2])), indexDestination=TRUE)
  
  return(outputname)
}
```

### Example Usage
```{r eval = FALSE}
l23 <- makesubclassBAM(
  inputdir="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Random100_PVALBvsL23_inVisp/l23_outputs/STAR_results/coord_bams", 
  outputname="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_random100_merged_l23.bam", chrID="NC_000082.7",
  coords=c(18038612,18056471))

pvalb <- makesubclassBAM(
  inputdir="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Random100_PVALBvsL23_inVisp/pvalb_outputs/STAR_results/coord_bams", 
  outputname="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_random100_merged_pvalb.bam", chrID="NC_000082.7",
  coords=c(18038612,18056471))
```

## Create Reference Genome Variables
```{r}
refseq_ref_GRCm38 <- ("/external/rprshnas01/netdata_kcni/stlab/Genomic_references/AIBS/Mouse/Refseq_GRCm38.p3/Raw/GCF_000001635.23_GRCm38.p3_genomic.gff")

refseq_ref_GRCm39 <- ("/external/rprshnas01/netdata_kcni/stlab/Genomic_references/Refseq/Mouse/Refseq_GRCm39/Raw/GCF_000001635.27_GRCm39_genomic.gff")
  
ensembl_ref_GRCm39 <- ("/external/rprshnas01/netdata_kcni/stlab/Genomic_references/Ensembl/Mouse/Release_104/Raw/Mus_musculus.GRCm39.104.gtf")
```

### Create TxDb Object
* These take a while, so you can just run the one you need

```{r results=FALSE, message=FALSE, warning=FALSE}
refseqTxDb_GRCm39 <- makeTxDbFromGFF(refseq_ref_GRCm39)
```

```{r}
#refseqTxDb_GRCm38 <- makeTxDbFromGFF(refseq_ref_GRCm38)
```

```{r}
#ensemblTxDb <- makeTxDbFromGFF(ensembl_ref)
```

## Set Up Figure
The function below will create a coverage + sashimi plot for two BAMs aligned to gene isoform models for the chosen reference, additionally showing coordinate positions on the x-axis

**Parameters:**

* **txdb:** a TxDb object, can be *refseqTxDb_GRCm39*, *refseqTxDb_GRCm38*, OR *ensemblTxDb*
* **bam1/bam2:** as a path to the BAM file (should be defined in section 1.3 above, e.g. pvalb)
* **chrID:** as a character, representing chromosome for selected reference, e.g. "NC_000082.6" for RefSeq OR "16" for Ensembl
* **coords:** as a vector, where the 1st value is the start coordinate and the 2nd is the end, e.g. c(0, 100)—note that unit is in bases, so 18,050kb becomes 18050000

```{r}
makefigure <- function(txdb, bam1, bam2, chrID, coords, bam1title=deparse(substitute(bam1)), bam2title=deparse(substitute(bam2))) {
  
  # Extract list of transcripts
  gene_tx <- data.frame(transcripts(txdb)) %>% 
    filter(seqnames==chrID) %>% 
    filter((start>=coords[1]) & (end<=coords[2])) %>% 
    dplyr::select(tx_name) %>% 
    list(.$tx_name) %>% 
    .[[2]]
  
  # Create genome axis track
  genomeAxis <- GenomeAxisTrack(name="MyAxis", col="lightsteelblue4", fontcolor="lightsteelblue4") 
  
  # Create BamFile objects (used in sashimi threshold)
  bf_bam1 <- BamFile(bam1, asMates=TRUE)
  bf_bam2 <- BamFile(bam2, asMates=TRUE)
  
  # Create plot for coverage and sashimi
  bam1_peakReads <- AlignmentsTrack(bam1, name=bam1title, cex=2, background.title="white",
                                    sashimiScore=(1/1200)*(countBam(bf_bam1)$records),
                                    col.axis="lightsteelblue4", col.title="lightsteelblue4")
  bam2_peakReads <- AlignmentsTrack(bam2, name=bam2title, cex=2, background.title="white",
                                    sashimiScore=(1/1200)*(countBam(bf_bam2)$records),
                                    col.axis="lightsteelblue4", col.title="lightsteelblue4")
  
  # Create gene models
  gr <- exonsBy(txdb, by = "tx", use.names=TRUE)[gene_tx]
  gr <- unlist(gr)
  elementMetadata(gr)$transcript <- names(gr)
  gene_models <- Gviz::GeneRegionTrack(gr, showId=TRUE, options(ucscChromosomeNames=FALSE),
                                       transcriptAnnotation="transcript", name="Gene Model",
                                       background.title="white", col.axis="lightsteelblue4",
                                       col.title="lightsteelblue4", fill="darkgrey",
                                       fontcolor.group="lightsteelblue4", col.line="lightsteelblue4")
  
  # Put the figure together
  options(ucscChromosomeNames=FALSE)
  fig <- plotTracks(c(genomeAxis, bam1_peakReads, bam2_peakReads, gene_models),
             showId=TRUE,
             transcriptAnnotation="transcript",
             chromosome=chrID,
             sizes=c(1,3,3,2),
             from=coords[1],
             to=coords[2],
             extend.left=3500,
             fill="lightsteelblue", col.sashimi="lightsteelblue4",
             type=c('coverage', 'sashimi'))

  return(fig)
}
```

# Visualization: gviz

## Create Figure
* May optionally save to .png

```{r message=FALSE, warning=FALSE, results=FALSE}
pvalb <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_random100_merged_pvalb.bam")
l23 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_random100_merged_l23.bam")

#png(filename="pvalbvsl23.png")
makefigure(refseqTxDb_GRCm39, pvalb, l23, "NC_000082.7", c(18038612,18056471), bam1title="Pvalb", bam2title="L2/3 IT CTX-1")
#dev.off()
```

# Quantification (still testing)

## See exon boundary coordinates
```{r eval = FALSE}
gene_tx <- data.frame(transcripts(refseqTxDb_GRCm39)) %>% 
  filter(seqnames=="NC_000082.7") %>% 
  filter((start>=18038612) & (end<=18056471)) %>% 
  dplyr::select(tx_name) %>% 
  list(.$tx_name) %>% 
  .[[2]]
gr <- exonsBy(refseqTxDb_GRCm39, by = "tx", use.names=TRUE)[gene_tx]
data.frame(unlist(gr)) %>% 
  select(start, end, exon_name) 
```

## Define transcript-specific ranges
```{r eval = FALSE}
NM_172151.5_specific <- GRanges("NC_000082.7", IRanges(18039568, 18041241))
XM_036159976.1_or_XM_006522224.5 <- GRanges("NC_000082.7", IRanges(18053017, 18056471))
XM_036159977.1_or_XM_006522224.5 <- GRanges("NC_000082.7", IRanges(18045219, 18045310))
rng <- NM_172151.5_specific
```

## summarizeOverlaps: calculate normalized reads

### As a percentage of reads for whole gene
```{r eval = FALSE}
#Pvalb cells
bf_pvalb <- BamFile(pvalb, asMates=TRUE)

counts_pvalb <- data.frame(union=assays(summarizeOverlaps(rng, bf_pvalb, singleEnd=FALSE, fragments=TRUE))$counts)
colSums(counts_pvalb) # raw number of reads in region
p_pvalb <- 100*(rowSums(counts_pvalb))/(countBam(bf_pvalb)$records) # reads as proportion of total reads in gene

#L2/3
bf_l23 <- BamFile(l23, asMates=TRUE)

counts_l23 <- data.frame(union=assays(summarizeOverlaps(rng, bf_l23, singleEnd=FALSE, fragments=TRUE))$counts)
colSums(counts_l23)
p_l23 <- 100*(rowSums(counts_l23))/(countBam(bf_l23)$records)

colSums(data.frame(p_pvalb))
colSums(data.frame(p_l23))
```

### Region Ratios (short vs long-minus-short)
```{r}
long <- GRanges("NC_000082.7", IRanges(18039568, 18041241))
short <- GRanges("NC_000082.7", IRanges(18038612, 18039568))

long_pvalb <- data.frame(union=assays(summarizeOverlaps(long, bf_pvalb, singleEnd=FALSE, fragments=TRUE))$counts)
short_pvalb <- data.frame(union=assays(summarizeOverlaps(short, bf_pvalb, singleEnd=FALSE, fragments=TRUE))$counts)

long_l23 <- data.frame(union=assays(summarizeOverlaps(long, bf_l23, singleEnd=FALSE, fragments=TRUE))$counts)
short_l23 <- data.frame(union=assays(summarizeOverlaps(short, bf_l23, singleEnd=FALSE, fragments=TRUE))$counts)

ratio_pvalb <- (rowSums(long_pvalb))/(rowSums(short_pvalb))
ratio_l23 <- (rowSums(long_l23))/(rowSums(short_l23))

colSums(data.frame(ratio_pvalb))
colSums(data.frame(ratio_l23))
```
### As a percentage of reads in long final exon region
```{r}
full <- GRanges("NC_000082.7", IRanges(18038617, 18041241))
full_minus_short <- GRanges("NC_000082.7", IRanges(18039568, 18041241))

full_pvalb <- data.frame(union=assays(summarizeOverlaps(full, bf_pvalb, singleEnd=FALSE, fragments=TRUE))$counts)
full_minus_short_pvalb <- data.frame(union=assays(summarizeOverlaps(full_minus_short, bf_pvalb, singleEnd=FALSE, fragments=TRUE))$counts)

full_l23 <- data.frame(union=assays(summarizeOverlaps(full, bf_l23, singleEnd=FALSE, fragments=TRUE))$counts)
full_minus_short_l23 <- data.frame(union=assays(summarizeOverlaps(full_minus_short, bf_l23, singleEnd=FALSE, fragments=TRUE))$counts)

p_pvalb <- 100*(rowSums(full_minus_short_pvalb))/(rowSums(full_pvalb))
p_l23 <- 100*(rowSums(full_minus_short_l23))/(rowSums(full_l23))

colSums(data.frame(p_pvalb))
colSums(data.frame(p_l23))
```
## summarizeJunctions
18041241-18042082 (long final exon) 
* Pvalb = 50
* L23: 16

18039568-18042082 (short final exon) 
* Pvalb = 8
* L23: 117

```{r}
galign_pvalb <- readGAlignmentPairs(pvalb,index="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_random100_merged_pvalb.bam", strandMode=1)
galign_l23 <- readGAlignmentPairs(l23, index="/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_random100_merged_l23.bam", strandMode=1)

junctions_pvalb <- summarizeJunctions(galign_pvalb)
junction_df_pvalb <- data.frame(c(as.data.frame(junctions_pvalb@ranges), as.data.frame(junctions_pvalb$score)))

junctions_l23 <- summarizeJunctions(galign_l23)
junction_df_l23 <- data.frame(c(as.data.frame(junctions_l23@ranges), as.data.frame(junctions_l23$score)))

junction_df_l23 %>% filter(end==18042081)
junction_df_pvalb %>% filter(end==18042081)
```

## featureCounts
```{r}
fc_pvalb <- featureCounts(files=pvalb, 
              annot.ext=refseq_ref_GRCm39,
              isGTFAnnotationFile = TRUE,
              GTF.attrType = "ID",
              isPairedEnd = TRUE,
              juncCounts = TRUE,
              genome = "/external/rprshnas01/netdata_kcni/stlab/Genomic_references/Refseq/Mouse/Refseq_GRCm39/Raw/GCF_000001635.27_GRCm39_genomic.fna")

fc_l23 <- featureCounts(files=l23, 
              annot.ext=refseq_ref_GRCm39,
              isGTFAnnotationFile = TRUE,
              GTF.attrType = "ID",
              isPairedEnd = TRUE,
              juncCounts = TRUE,
              genome = "/external/rprshnas01/netdata_kcni/stlab/Genomic_references/Refseq/Mouse/Refseq_GRCm39/Raw/GCF_000001635.27_GRCm39_genomic.fna")
```
