---
title: 'Isoform Project: Visualizations'
output: html_document
---

# 1. Set Up

## 1.1. Load Packages
```{r message=FALSE, warning=FALSE}
library(Gviz)
library(rtracklayer)
library(Rsamtools)
library(GenomicFeatures)
library(ggbio)
library(GenomicRanges)
library(rtracklayer)
library(Rsamtools)
library(GenomicAlignments)
library(VariantAnnotation)
library(ggforce)
library(tidyverse)
```

## 1.2. Choose RefSeq or Ensembl
* RefSeq = 1
* Ensembl = 2

```{r}
method <- 1
```

## 1.3. Load BAMs

```{r}
#RefSeq -------------------------------------------------------------------------

if(method==1){
  
  #Testing
  WT1 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_tests/RefseqForAIBSSumOverlapsScript_RamseyTestFile/WT1Aligned.sortedByCoord.out.bam")


  #Pvalb
  ls15581 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15581_S09_E1-50_R1Aligned.sortedByCoord.out.bam")

  smdaifd <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-DAIFD_S51_E1-50_R1Aligned.sortedByCoord.out.bam")

  smd9cxp <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-D9CXP_S75_E1-50_R1Aligned.sortedByCoord.out.bam")

  ls15542 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15542_S58_E1-50_R1Aligned.sortedByCoord.out.bam")

  ls15074 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15074_S67_E1-50_R1Aligned.sortedByCoord.out.bam")


  #L2/3
  ls15380 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15380_S42_E1-50_R1Aligned.sortedByCoord.out.bam")

  smd9e61 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-D9E61_S93_E1-50_R1Aligned.sortedByCoord.out.bam")

  smd9eqh <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-D9EQH_S77_E1-50_R1Aligned.sortedByCoord.out.bam")

  ls15379 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15379_S07_E1-50_R1Aligned.sortedByCoord.out.bam")

  ls15060 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15060_S30_E1-50_R1Aligned.sortedByCoord.out.bam")

}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  ls15581<-("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Ensembl_Pilot/LS-15581_S09_E1-50_Aligned.sortedByCoord.out.bam")

  smdaifd <-("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Ensembl_Pilot/SM-DAIFD_S51_E1-50_Aligned.sortedByCoord.out.bam")

  ensembl_ls15379 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Ensembl_untarred_outputs/Aligned.sortedByCoord.out.bam")

}
```

### 1.3.1. Create Index for BAM
**Only if no .bai file for BAM has been created yet**
```{r}
#indexBam()
```

## 1.4. Load Reference Genome
```{r}
#RefSeq -------------------------------------------------------------------------

if(method==1){
  
  refseq_ref <- ("/external/rprshnas01/netdata_kcni/stlab/Genomic_references/AIBS/Mouse/Refseq_GRCm38.p3/Raw/GCF_000001635.23_GRCm38.p3_genomic.gff")

}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  ensembl_ref <- ("/external/rprshnas01/netdata_kcni/stlab/Genomic_references/Ensembl/Mouse/Release_104/Raw/Mus_musculus.GRCm39.104.gtf")
  
}
```

## 1.5. Create TxDb

```{r  message=FALSE, warning=FALSE}
#RefSeq -------------------------------------------------------------------------

if(method==1){

  refseqTxDb <- makeTxDbFromGFF(refseq_ref)

}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  ensemblTxDb <- makeTxDbFromGFF(ensembl_ref)
  
}
```

## 1.6. Chromosome Identifier

```{r}
#RefSeq -------------------------------------------------------------------------

if(method==1){
  
  chrID <- "NC_000082.6"  #Chromosome 16

}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  chrID <- "16"  #Chromosome 16
  
}
```

##1.7. Gene Coordinates
* Coordinates below are for zdhhc8.

```{r}
#RefSeq -------------------------------------------------------------------------

if(method==1){
  
  gene_start <- 18220748
  gene_stop <- 18235153
  
}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  gene_start <- 18038617
  gene_stop <- 18053000
  
}
```

## 1.8. Get list of transcripts for gene of interest

### 1.8.1. Create dataframe of all transcripts in genome reference

```{r}
#RefSeq -------------------------------------------------------------------------

if(method==1){

  refseqTxDb_df <- data.frame(transcripts(refseqTxDb))
  
}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  ensemblTxDb_df <- data.frame(transcripts(ensemblTxDb))
  
}
```

### 1.8.2. Create list of transcripts located within gene coordinates

```{r}
#RefSeq -------------------------------------------------------------------------

if(method==1){
  
  gene_df <- refseqTxDb_df %>% 
    filter(seqnames==chrID) %>% 
    filter((start>=gene_start) & (end<=gene_stop)) %>% 
    dplyr::select(tx_name)

  gene_tx <- list(gene_df$tx_name)
  gene_tx <- gene_tx[[1]]
  
}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  gene_df <- ensemblTxDb_df %>% 
    filter(seqnames==chrID) %>% 
    filter((start>=gene_start) & (end<=gene_stop)) %>% 
    dplyr::select(tx_name)

  gene_tx <- list(gene_df$tx_name)
  gene_tx <- gene_tx[[1]]
  
}
```

# 2. Visualization

## 2.1. Genome Axis Track
* This will be used as an x-axis to indicate genomic coordinates for other plots.

```{r}
genomeAxis <- GenomeAxisTrack(name="MyAxis") 
```

## 2.2. Gene Models
* This will be used to display the different gene isoforms aligned to the gene of interest.

```{r message=FALSE, warning=FALSE}
#RefSeq -------------------------------------------------------------------------

if(method==1){
  
  gr <- exonsBy(refseqTxDb, by = "tx", use.names=TRUE)[gene_tx]
  gr <- unlist(gr)
  elementMetadata(gr)$transcript <- names(gr)
  gene_models <- Gviz::GeneRegionTrack(gr, showId=TRUE, options(ucscChromosomeNames=FALSE),
                                       transcriptAnnotation="transcript")

}

#Ensembl -------------------------------------------------------------------------

if(method==2){
  
  gr <- exonsBy(ensemblTxDb, by = "tx", use.names=TRUE)[gene_tx]
  gr <- unlist(gr)
  elementMetadata(gr)$transcript <- names(gr)
  gene_models <- Gviz::GeneRegionTrack(gr, showId=TRUE, options(ucscChromosomeNames=FALSE),
                                       transcriptAnnotation="transcript")

}
```

## 2.3. Alignments Track for BAM of interest
* This will be used for the coverage + sashimi plot.

```{r}
peakReads <- AlignmentsTrack(smd9eqh) #this is one of the BAM files loaded in beginning 
```

## 2.4. Create Plot
* Includes coverage, sashimi, and transcript models
* May optionally save image to .png

```{r}
#png(filename="ls15060.png")
options(ucscChromosomeNames=FALSE)
plotTracks(c(peakReads, gene_models, genomeAxis),
           showId=TRUE,
           transcriptAnnotation="transcript",
           chromosome=chrID,
           sizes=c(3,2,1),
           from=gene_start,
           to=gene_stop,
           extend.left=3500,
           fill="lightsteelblue", col.sashimi="lightsteelblue4",
           title.width=1.5,
           type=c('coverage', 'sashimi'))
#dev.off()
```