---
title: 'Isoform Project: Visualizations'
output: html_document
---

# **1.** Set Up

## **1.1.** Load Packages
```{r message=FALSE, warning=FALSE}
library(Gviz)
library(rtracklayer)
library(Rsamtools)
library(GenomicFeatures)
library(ggbio)
library(GenomicRanges)
library(rtracklayer)
library(Rsamtools)
library(GenomicAlignments)
library(VariantAnnotation)
library(ggforce)
library(tidyverse)
```

## **1.2.** Create BAM Variables
```{r}
#RefSeq -------------------------------------------------------------------------


#Testing (Bulk)
WT1 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/RefSeq_tests/RefseqForAIBSSumOverlapsScript_RamseyTestFile/WT1Aligned.sortedByCoord.out.bam")


#Pvalb
ls15581 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15581_S09_E1-50_R1Aligned.sortedByCoord.out.bam")

smdaifd <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-DAIFD_S51_E1-50_R1Aligned.sortedByCoord.out.bam")

smd9cxp <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-D9CXP_S75_E1-50_R1Aligned.sortedByCoord.out.bam")

ls15542 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15542_S58_E1-50_R1Aligned.sortedByCoord.out.bam")

ls15074 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15074_S67_E1-50_R1Aligned.sortedByCoord.out.bam")


#L2/3
ls15380 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15380_S42_E1-50_R1Aligned.sortedByCoord.out.bam")

smd9e61 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-D9E61_S93_E1-50_R1Aligned.sortedByCoord.out.bam")

smd9eqh <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/SM-D9EQH_S77_E1-50_R1Aligned.sortedByCoord.out.bam")

ls15379 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15379_S07_E1-50_R1Aligned.sortedByCoord.out.bam")

ls15060 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Pilot_PVALBvsL23_inVisp/Pilot_coord_bams/STAR_results/coord_bams/LS-15060_S30_E1-50_R1Aligned.sortedByCoord.out.bam")


#Ensembl -------------------------------------------------------------------------

ensembl_ls15379 <- ("/external/rprshnas01/netdata_kcni/stlab/Intralab_collab_scc_projects/Isoform_project/Ensembl_untarred_outputs/Aligned.sortedByCoord.out.bam")

```

### **1.2.1.** Create Index for BAMs
**Only if no .bai file for BAM has been created yet**
```{r}
#indexBam()
```

## **1.3.** Create Reference Genome Variables
```{r}
refseq_ref <- ("/external/rprshnas01/netdata_kcni/stlab/Genomic_references/AIBS/Mouse/Refseq_GRCm38.p3/Raw/GCF_000001635.23_GRCm38.p3_genomic.gff")
  
ensembl_ref <- ("/external/rprshnas01/netdata_kcni/stlab/Genomic_references/Ensembl/Mouse/Release_104/Raw/Mus_musculus.GRCm39.104.gtf")
```

### **1.3.1** Create TxDb Object
* These take a while, so you can just run the one you need

```{r}
refseqTxDb <- makeTxDbFromGFF(refseq_ref)
```

```{r}
#ensemblTxDb <- makeTxDbFromGFF(ensembl_ref)
```

## **1.4.** Set Up Figure
The function below will create a coverage + sashimi plot for the chosen BAM aligned to gene isoform models for the chosen reference, additionally showing coordinate positions on the x-axis

**Parameters:**

* txdb: a TxDb object, can be either **refseqTxDb** OR **ensemblTxDb**
* bam: as a variable (should be defined in section 1.2 above), e.g. ls15581
* chrID: as a character representing chromosome for selected reference, e.g. "NC_000082.6" for RefSeq OR "16" for Ensembl
* coords: as a vector where the 1st value is the start coordinate and the 2nd is the end, e.g. c(0, 100)

```{r}
makefigure <- function(txdb, bam, chrID, coords) {
  
  # Name the coordinate parameters
  gene_start <- coords[1]
  gene_stop <- coords[2]
  
  # Extract list of transcripts
  gene_tx <- data.frame(transcripts(txdb)) %>% 
    filter(seqnames==chrID) %>% 
    filter((start>=gene_start) & (end<=gene_stop)) %>% 
    dplyr::select(tx_name) %>% 
    list(.$tx_name) %>% 
    .[[2]]
  
  # Create genome axis track
  genomeAxis <- GenomeAxisTrack(name="MyAxis") 
  
  # Create plot for coverage and sashimi
  peakReads <- AlignmentsTrack(bam)
  
  # Create gene models
  gr <- exonsBy(txdb, by = "tx", use.names=TRUE)[gene_tx]
  gr <- unlist(gr)
  elementMetadata(gr)$transcript <- names(gr)
  gene_models <- Gviz::GeneRegionTrack(gr, showId=TRUE, options(ucscChromosomeNames=FALSE),
                                       transcriptAnnotation="transcript")
  
  # Put the figure together
  options(ucscChromosomeNames=FALSE)
  fig <- plotTracks(c(peakReads, gene_models, genomeAxis),
             showId=TRUE,
             transcriptAnnotation="transcript",
             chromosome=chrID,
             sizes=c(3,2,1),
             from=gene_start,
             to=gene_stop,
             extend.left=3500,
             fill="lightsteelblue", col.sashimi="lightsteelblue4",
             title.width=1.5,
             type=c('coverage', 'sashimi'))

  return(fig)
}
```

# **2.** Visualization

## **2.1** Create Figure
* May optionally save to .png

```{r message=FALSE, warning=FALSE, results=FALSE}
#png(filename="ls15581.png")
makefigure(refseqTxDb, ls15581, "NC_000082.6", c(18220748,18235153))
#dev.off()
```
